### PR TITLE
Compacter doesn't delete parent states for tipsets with height > main…

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
… branch height

### Description of the Change
Check height > main branch height and save them in compacter. This allows restart after long pause and compaction.

### CODE REVIEW

File: core/storage/compacter/compacter.cpp [hunk 1]
```diff
@@ -244,11 +247,14 @@ namespace fc::storage::compacter {
           primitives::tipset::put(nullptr, put_block_header, block);
           queue->push(*asBlake(block.messages));
         }
+        // if cache has tipsets with height > main_height, save them
+        auto tipset_res{interpreter_cache->tryGet(ts->key)};
+        if (tipset_res.has_value() && (ts->height() > head->height())) {
+          queue->push(*asBlake(tipset_res->value().state_root));
+          queue->push(*asBlake(tipset_res->value().message_receipts));
+        }
```
Comment (turuslan) [inline core/storage/compacter/compacter.cpp:255]:
```suggestion
        if (ts->height() >= head->height()) {
          if (auto res{interpreter_cache->tryGet(ts->key)}; res && *res) {
            queue->push(*asBlake(res->value().state_root));
            queue->push(*asBlake(res->value().message_receipts));
          }
        }
```

**Review by turuslan — APPROVED** [review]

File: core/storage/compacter/compacter.cpp [hunk 2]
```diff
@@ -234,6 +234,9 @@ namespace fc::storage::compacter {
       }
       ts = ts_load->load(ts->getParents()).value();
     }
+    auto head_res{interpreter_cache->get(head->key).value()};
```
Comment (wer1st) [inline core/storage/compacter/compacter.cpp:237]:
```suggestion
    const auto head_res{interpreter_cache->get(head->key).value()};
```

Comment (wer1st) [inline core/storage/compacter/compacter.cpp:237]: Could it has an error?

File: test/core/vm/interpreter/interpreter_cache_test.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    Result result{.state_root = state_root,
+                  .message_receipts = "010001020003"_cid,
+                  .weight = 1};
+    TipsetKey tipset_key;
+    interpreter_cache->set(tipset_key, result);
+
+    EXPECT_CALL(*ipld, get(Eq(asBlake(state_root)), Eq(nullptr)))
+        .WillOnce(Return(false));
+
+    auto res = interpreter_cache->tryGet(tipset_key);
```
Comment (wer1st) [inline test/core/vm/interpreter/interpreter_cache_test.cpp:35]:
```suggestion
    const auto res = interpreter_cache->tryGet(tipset_key);
```

**Review by wer1st — APPROVED** [review]

File: core/storage/compacter/compacter.cpp (hunk 2 above)
Comment (Alexey-N-Chernyshov) [reply core/storage/compacter/compacter.cpp:237]: No, chain head must be interpreted and cached.

Comment (wer1st) [reply core/storage/compacter/compacter.cpp]: ok

Comment (turuslan) [reply core/storage/compacter/compacter.cpp]: No, program keeps `head` consistent/correct, so there should not be error

**Pull request merged**